pkg/instance/instance.go

```diff
@@ -406,122 +413,205 @@ func (s *Server) spdkInstanceLog(req *rpc.InstanceLogRequest, srv rpc.InstanceSe
 	return grpcstatus.Error(grpccodes.Unimplemented, "spdk instance log is not supported")
 }
 
+func (s *Server) handleNotify(ctx context.Context, notifyChan chan struct{}, srv rpc.InstanceService_InstanceWatchServer) error {
+	logrus.Info("Start handling notify")
+
+	for {
+		select {
+		case <-ctx.Done():
+			logrus.Info("Stopped handling notify")
+			return ctx.Err()
+		case <-notifyChan:
+			if err := srv.Send(&empty.Empty{}); err != nil {
+				return errors.Wrap(err, "failed to send instance response")
+			}
+		}
+	}
+}
+
 func (s *Server) InstanceWatch(req *empty.Empty, srv rpc.InstanceService_InstanceWatchServer) error {
 	logrus.Info("Start watching instances")
 
+	done := make(chan struct{})
+
+	clients := map[string]interface{}{}
+	go func() {
+		<-done
+
+		logrus.Info("Stopped clients")
+		for name, c := range clients {
+			switch c := c.(type) {
+			case *client.ProcessManagerClient:
+				c.Close()
+			case *spdkclient.SPDKClient:
+				c.Close()
+			}
+			delete(clients, name)
+		}
+		close(done)
+	}()
+
+	// Create a client for watching processes
+	pmClient, err := client.NewProcessManagerClient("tcp://"+s.processManagerServiceAddress, nil)
+	if err != nil {
+		done <- struct{}{}
+		return grpcstatus.Error(grpccodes.Internal, errors.Wrapf(err, "failed to create ProcessManagerClient").Error())
+	}
```

Comment:
The error message doesn't include the context of the instance. Just noticed most of the logs here have no context, so it will cause logs of each instance monitor same and are unable to differentiate. Let's create a ticket for later improvement.

Reply:
Tracked in https://github.com/longhorn/longhorn/issues/6724